### ISSUE TITLE
`use_key_in_widget_constructors` should report only if super has a constructor with `key`

### ISSUE THREAD

**Issue by asashour** [dart-lang/sdk#58727]
case 1:
```dart
import 'package:flutter/material.dart';

class ParentWidget extends StatelessWidget { // the lint here is ok
  @override
  Widget build(BuildContext context) => Container();
}

class MyWidget extends ParentWidget { // <--- the lint here is wrong
}
```

case 2:
```dart
import 'package:flutter/material.dart';

class ParentWidget extends StatelessWidget { // the lint here is ok
  int? i;

  ParentWidget({this.i});

  @override
  Widget build(BuildContext context) => Container();
}

class MyWidget extends ParentWidget { // <--- the lint here is wrong
}
```
case 3:
```dart
import 'package:flutter/material.dart';

class ParentWidget extends StatelessWidget {
  ParentWidget({String key = 'abc'});

  @override
  Widget build(BuildContext context) => Container();
}

class MyWidget extends ParentWidget {
}
```
**Expected behavior**
In `MyWidget`, there shouldn't be a lint, because the parent constructor doesn't have a `key` with type `Key`.

**Actual behavior**
In `MyWidget`, there is a lint reported.

**Comment by FMorschel** [comment]
FYI @Varadraj75

**Comment by Varadraj75** [comment]
I just did the quick fix side of this in #63999, so this one's a natural fit. Reproducing locally now, happy to take it if it's free.

**Comment by FMorschel** [comment]
I don't think anyone is working on this. Thanks for offering!

**Comment by Varadraj75** [comment]
Reproduced all three cases. The rule already skips reporting when an explicit super(...) call targets a constructor with no Key typed key parameter, but the no-constructor and implicit super paths never check the super at all, so MyWidget() : super(); doesn't lint today while the same class without a constructor does. Plan is to give those paths that same existing check, which fixes all three cases here. Can condition it on the superclass being from another package instead if that's preferred.

**Comment by bwilkerson** [comment]
> Can condition it on the superclass being from another package instead if that's preferred.

Are you thinking of the case (from above) of a class that the user can't update?

If so, the check would probably be more subtle than just having the superclass be in a different package. Many projects have multiple packages, all of which are under the user's control.

**Comment by Varadraj75** [comment]
Yeah, that was about the can't-update case, and you're right that package boundaries don't really capture it. This PR doesn't do any of that, it just gives the implicit-super and no-constructor paths the same super constructor check the explicit super(...) path already has, so the three cases in the issue stop reporting. Whether to keep reporting when the user does control the superclass feels like a separate call, happy to follow up if you want to go that way.

**Comment by bwilkerson** [comment]
Thanks, both for the work already done and the offer. Unless we have evidence that this is impacting a reasonable number of users I wouldn't consider the latter case to be urgent.

**Comment by Varadraj75** [comment]
@FMorschel  @bwilkerson 

This issue is landed in https://github.com/dart-lang/sdk/pull/64104 

We can close it.
